src/foam/nanos/crunch/UserCapabilityJunctionRefine.js

```diff
@@ -169,6 +179,56 @@ foam.CLASS({
 
         return objectToSave;
       `
+    },
+    {
+      name: 'getSubject',
+      type: 'foam.nanos.auth.Subject',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `
+        UserCapabilityJunction ucj = this;
+        var currentSubject = (Subject) x.get("subject");
+        var userDAO = (DAO) x.get("userDAO");
+
+        Subject subject = new Subject(x);
+        if ( ucj instanceof AgentCapabilityJunction ) {
+          subject.setUser((User) userDAO.find(ucj.getSourceId()));
+          AgentCapabilityJunction acj = (AgentCapabilityJunction) ucj;
+          subject.setUser((User) userDAO.find(acj.getEffectiveUser()));
+          return subject;
+        }
+
+        // We will need the capability object to know how it's associated
+        var capabilityDAO = (DAO) x.get("capabilityDAO");
+        var cap = (Capability) capabilityDAO.find(ucj.getTargetId());
+        if ( cap == null ) {
+          throw new RuntimeException(
+            "Tried to call getSubject() on UCJ with unrecognized capability");
+        }
+
+        if ( ucj.getSourceId() == currentSubject.getUser().getId() ) {
+          subject.setUser(currentSubject.getRealUser());
+          subject.setUser(currentSubject.getUser());
+          return subject;
+        }
+
+        if ( cap.getAssociatedEntity() == USER ) {
+          subject.setUser((User) userDAO.find(
+            0 != ucj.getLastUpdatedRealUser()
+              ? ucj.getLastUpdatedRealUser()
+              : ucj.getSourceId()
+          ));
```

Comment:
suggest remove:
          subject.setUser((User) userDAO.find(
            0 != ucj.getLastUpdatedRealUser()
              ? ucj.getLastUpdatedRealUser()
              : ucj.getSourceId()
          ));

replace with:
          if ( ucj.getLastUpdatedRealUser() != 0 ) {
            subject.setUser((User) userDAO.find(ucj.getLastUpdatedRealUser());
          }

This suggestion ensures your not setting both subject.user and subject.realUser as the same user. Feel like that might conflict if anyone is checking if both entities are there.

Further I wonder... getLastUdpatedRealUser() ...

So we are in the situation where ucj.sourceId is a business, and lastUpdatedRealUser is someone in this business or potentially a system user id. 

question - what if the last person to update this ucj is not related to the business?

Reply:
The intention is that this is only set when the capability will be granted by an approval but you still need to know something about the `realUser` when it's granted.

Specifically for NP, the condition we have to auto-grant the review capabilities requires information about this user, but it can only happen after approval has been granted.

Reply:
In other words what we want is "what user last changed this capability as the business - do their privileges imply a review?" I haven't thought of a better solution to this problem yet.